### ISSUE TITLE
bulk_save_object inserts new objects twice

### ISSUE THREAD

**Issue by eric-nieuwland** [sqlalchemy/sqlalchemy#5833]
**Describe the bug**
`bulk_save_object` inserts new objects with foreign key relations twice in the database when called without `return_defaults`.

**Expected behavior**
Each object should be inserted only once, possibly without updating the foreign keys.

**To Reproduce**
Entity A contains some data, e.g. a string, and an explicit `_id = Column(Integer, primary_key=True)`.
Entity B contains a reference to Entity A: `Column(Integer, ForeignKey('entity_a._id'), index=True)`.
Create and commit an instance of Entity A.
Create an instance of Entity B, which references the Entity A instance.

Execute `session. bulk_save_object([b])` and commit; now check the database.
Table for Entity B now contains 2 rows:
 - with NULL value as reference
 - with Entity A _id value

Execute `session. bulk_save_object([b], return_defaults=True)` and commit; now check the database.
Table for Entity B now contains 1 row:
 - with Entity A _id value

**Versions.**
 - OS: MacOS 11.1 and Windows 10
 - Python: 3.9.1
 - SQLAlchemy: 1.3.22
 - Database: sqlite

**Additional context**

**Have a nice day!**

**Comment by zzzeek** [comment]
hi there-

there's no way I'm going to be able to understand what you're doing without a runnable [MCVE](http://stackoverflow.com/help/mcve).  there's no mechanism I'm aware of in the "bulk save" that could insert an object "twice" and what you are observing is likely highly dependent on the specific mappings in use as well as the data and means of calling upon the API.    

thanks!

**Comment by eric-nieuwland** [comment]
```py
import pathlib

import sqlalchemy
import sqlalchemy.orm
from sqlalchemy.ext.declarative import declarative_base


Base = declarative_base()


class EntityA(Base):
    __tablename__ = "entity_a"

    _id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
    name = sqlalchemy.Column(sqlalchemy.String)

    bs = sqlalchemy.orm.relationship("EntityB", back_populates="a")

    def __init__(self, name):
        self.name = name

    def __repr__(self):
        return f"<EntityA {self._id} '{self.name}'>"


class EntityB(Base):
    __tablename__ = "entity_b"

    _id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
    _a_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('entity_a._id'))

    a = sqlalchemy.orm.relationship("EntityA", back_populates="bs")

    def __init__(self, a):
        self.a = a

    def __repr__(self):
        return f"<EntityB {self._id} {self._a_id} {self.a!r}>"


db_file = pathlib.Path(__file__).resolve().with_suffix(".db")
db_file.unlink(missing_ok=True)
db_engine = sqlalchemy.create_engine(f"sqlite:///{db_file}")
Base.metadata.create_all(db_engine)

db_session = sqlalchemy.orm.sessionmaker(db_engine)()

a = EntityA("foo")
db_session.add(a)
db_session.commit()

b = EntityB(a)
db_session.bulk_save_objects([b])
db_session.commit()

for obj in db_session.query(EntityB):
    print(repr(obj))
```

**Comment by zzzeek** [comment]
hey there -

yeah, "EntityB" is being added to the session, in addition to your entering it into bulk_save_objects(), so that's why you see it twice.

EntityB gets added right here:

    b = EntityB(a)

because the now-deprecated default behavior is that the backref of EntityA.bs will cascade the B objects into the Session.

Here's links to background on this:

https://docs.sqlalchemy.org/en/13/orm/cascades.html#controlling-cascade-on-backrefs

https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#cascade-backrefs-behavior-deprecated-for-removal-in-2-0

If I run your program in 1.4 with [2.0 deprecations mode](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#sqlalchemy-2-0-deprecations-mode) turned on, the behavior is apparent:

```
$ SQLALCHEMY_WARN_20=1 python -W always test3.py 
test3.py:8: MovedIn20Warning: The ``declarative_base()`` function is now available as sqlalchemy.orm.declarative_base(). (deprecated since: 1.4) (Background on SQLAlchemy 2.0 at: http://sqlalche.me/e/b8d9)
  Base = declarative_base()
/home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/unitofwork.py:65: RemovedIn20Warning: "EntityB" object is being merged into a Session along the backref cascade path for relationship "EntityA.bs"; in SQLAlchemy 2.0, this reverse cascade will not take place.  Set cascade_backrefs to False for the 2.0 behavior; or to set globally for the whole Session, set the future=True flag (Background on SQLAlchemy 2.0 at: http://sqlalche.me/e/b8d9)
  _warn_for_cascade_backrefs(item_state, prop)
<EntityB 1 None None>
<EntityB 2 1 <EntityA 1 'foo'>>

```

so it's a little confusing for now but help is on the way!

**Comment by eric-nieuwland** [comment]
Thanks! That will teach me to read the manual more closely ;-)
You may want to add a note to the description of `bulk_save_objects`.

**Comment by zzzeek** [comment]
the confusion here IMO is mostly over the non-intuitive cascade behavior which we are removing so I think this issue will resolve itself.
